### PR TITLE
fix(permit): presign

### PR DESCRIPTION
**Pull request by alfetopito**
## The problem

Orders that did not need permit hooks included it in their appData.
The hook included was however for the permit signer account rather than the user account.
This is bad in 2 ways:
1. User is paying for the approval unnecessarily.
2. It break the gas estimation for users trying to approve the same token later on

For a full picture of how often and how many times it happened, check [this dune query](https://dune.com/queries/3376950?category=abstraction&namespace=cow_protocol&table=trades&blockchains=ethereum)

image

## Type of orders

Analyzing only the most recent cases, there are [2 approvals](https://etherscan.io/tokenapprovalchecker?search=0xce69d355dfdf13c3ead95ec1c437df5d4bac05e4)

The [first](https://etherscan.io/tx/0x5e8464cccf089aa85b6ac4a03ea53f45c31092631a305108911baa0df399c97d) approves DAI in [this order](https://explorer.cow.fi/orders/0x6147b0fe9a3d83f9473e55a11e189056cd6eb8863dfee9c26b5d302aa5e4fc2515ecfc5b05f1db628f2cb7954cf9384146f3a18065a7b4b2?tab=overview).
The [second](https://etherscan.io/tx/0xb754d2db5fce616813ddfd632701bd360ef34951b8cab367d382449be55e153e) approves XAI in [this order](https://explorer.cow.fi/orders/0x62471978efbc349f9083dfcf43627d1e7f1360dcb923434868bc3a40a633f80781310f3aa1c1424d1bef20f8848d4f69948afe1265abbd85?tab=overview)

They both are pre-sign orders and both are Safes.

Curiously enough, both orders were placed as part of a Safe Approval+pre-sign bundle. See [DAI tx](https://app.safe.global/transactions/tx?safe=eth:0x15ecfc5b05f1db628f2cb7954cf9384146f3a180&id=multisig_0x15ecfC5b05F1db628f2cb7954Cf9384146f3A180_0xe8754328f55985ad02e9b705635be8c0231f63236772c85eaf088b543dbd73a2) and [XAI tx](https://app.safe.global/transactions/tx?safe=eth:0x81310f3AA1C1424D1Bef20f8848d4F69948aFE12&id=multisig_0x81310f3AA1C1424D1Bef20f8848d4F69948aFE12_0xb631bb28dc62f6e53fad2e80400955d387e20d6404e25babf84706e53ce62626).

## Conclusion

[Safe Bundle flow](https://github.com/cowprotocol/cowswap/blob/f6f4e4d0b143c4ecb056269facf2f26bce638b19/apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleApprovalFlow.ts#L1-L0) is passing the appData raw as it receives from the quote.

## To reproduce

Keep in mind I wasn't able to reproduce, but this is the path that seems to result in the bug.
So there might be something else causing this to sometimes include the bad appData.

1. Using a Safe, pick a token which is permittable but for which you have no approvals
2. Without signing the Safe tx, place the order and leave it in `signing` status
⚠️ It's important to NOT sign/execute it as we don't want screw up the signing permit account again
4. Open the order on Explorer
* It'll contain pre-hook in its appData - it SHOULDN'T!!

## Root problem

First of all, no SC flow should include pre-hooks at this point.
But it's just a symptom of another problem.

The flows usually only pass along the appData used for the quote.

In the example of the executed XAI approval, the quote [did contain the pre-hooks](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/r/s/bwYxf):

```
2024-01-20T12:02:46.175Z DEBUG request{id="522bf3eee62589be5b584fd3e3e81e18"}: shared::order_quoting: finished computing quote response=OrderQuoteResponse { quote: OrderQuote { sell_token: 0xd7c9f0e536dc865ae858b0c0453fe76d13c3beac, buy_token: 0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48, receiver: Some(0x81310f3aa1c1424d1bef20f8848d4f69948afe12), sell_amount: 989047090014311862272, buy_amount: 984754625, valid_to: 1705753960, app_data: Both { full: "{\"appCode\":\"CoW Swap-SafeApp\",\"environment\":\"production\",\"metadata\":{\"hooks\":{\"pre\":[{\"callData\":\"0xd505accf000000000000000000000000ce69d355dfdf13c3ead95ec1c437df5d4bac05e4000000000000000000000000c92e8bdf79f0507f65a392b0ab4667716bfe0110ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000000000000000000000000000000000006f135db7000000000000000000000000000000000000000000000000000000000000001c3eb01b8d580e5d4f663ce659b1e567897b96cf720d18c7ce1d110469bad28df17f23458d4eae1fc5c69e074da24923349e0eeaac3e9e6eda322181419b81e166\",\"gasLimit\":\"82438\",\"target\":\"0xd7C9F0e536dC865Ae858b0C0453Fe76D13c3bEAc\"}]},\"orderClass\":{\"orderClass\":\"market\"},\"quote\":{\"slippageBips\":\"50\"}},\"version\":\"0.11.0\"}", expected: 0xf3b711027604e0c67c83ec6d3dc9abb5d70e1feb47480c5ef35db2e50790e3de }, fee_amount: 10952909985688137728, kind: Sell, partially_fillable: false, sell_token_balance: Erc20, buy_token_balance: Erc20, signing_scheme: Eip712 }, from: 0x81310f3aa1c1424d1bef20f8848d4f69948afe12, expiration: 2024-01-20T12:04:41.183634010Z, id: Some(381631359) }
```

As expected, the exact case of the [DAI approval](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/r/s/t2SDO):

```
2024-01-17T10:35:30.295Z DEBUG request{id="d719ea183f2be678b793eb636ebe411b"}: shared::order_quoting: finished computing quote response=OrderQuoteResponse { quote: OrderQuote { sell_token: 0x6b175474e89094c44da98b954eedeac495271d0f, buy_token: 0x1f9840a85d5af5bf1d1762f925bdaddc4201f984, receiver: Some(0x15ecfc5b05f1db628f2cb7954cf9384146f3a180), sell_amount: 19975406096221803458560, buy_amount: 2950870380506139445207, valid_to: 1705489526, app_data: Both { full: "{\"appCode\":\"CoW Swap-SafeApp\",\"environment\":\"production\",\"metadata\":{\"hooks\":{\"pre\":[{\"callData\":\"0x8fcbaf0c000000000000000000000000ce69d355dfdf13c3ead95ec1c437df5d4bac05e4000000000000000000000000c92e8bdf79f0507f65a392b0ab4667716bfe01100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006f0f54740000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001c4eb8150c2464462b1127c53facfd301eb99eae8653693181436229c77d5c95874f21e0dd7ecbe03e1a269fd1aef79e59b6c3aa5994acce2265a82cd476c21a6f\",\"gasLimit\":\"84022\",\"target\":\"0x6B175474E89094C44Da98b954EedeAC495271d0F\"}]},\"orderClass\":{\"orderClass\":\"market\"},\"quote\":{\"slippageBips\":\"50\"}},\"version\":\"0.11.0\"}", expected: 0xe1057c3ef15694050d33992a6dfdd5a9cc99f4217ff075893a29ea9e77b2123f }, fee_amount: 24593903778196541440, kind: Sell, partially_fillable: false, sell_token_balance: Erc20, buy_token_balance: Erc20, signing_scheme: Eip712 }, from: 0x15ecfc5b05f1db628f2cb7954cf9384146f3a180, expiration: 2024-01-17T10:37:27.024912203Z, id: Some(378300082) }
```

## To reproduce

I was also not able to reproduce this.
Still unsure what's causing it.

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts [hunk 1]
```diff
@@ -46,7 +50,7 @@ export function AppDataHooksUpdater(): null {
       preInteractionHooks: permitData ? [permitData] : undefined,
     })
 
-    if (hooks) {
+    if (!isSmartContractWallet && hooks) {
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:53]: Maybe this solves the quote part

File: apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleApprovalFlow.ts [hunk 2]
```diff
@@ -54,6 +57,13 @@ export async function safeBundleApprovalFlow(
       amountToApprove: context.trade.inputAmount,
     })
 
+    // make sure no pre-approval hooks are included, just as a safety measure
+    if (appDataContainsHooks(orderParams.appData.fullAppData)) {
+      reportAppDataWithHooks(orderParams)
+      // wipe out the hooks
+      orderParams.appData = await updateHooksOnAppData(orderParams.appData, undefined)
+    }
+
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleApprovalFlow.ts:66]: If it's still quoted, as a last resort, remove if from the order params before it's signed.

File: libs/wallet/src/web3-react/hooks/useIsSmartContractWallet.ts [hunk 3]
```diff
@@ -41,14 +41,15 @@ export function useIsSmartContractWallet(): boolean {
...
   const isSmartContract = useCheckIsSmartContract()
   const isAmbireWallet = useIsAmbireWallet()
+  const isSafeWallet = useIsSafeWallet()
 
   useEffect(() => {
     if (!account) {
       setIsSmartContractWallet(false)
     } else {
-      setIsSmartContractWallet(Boolean(isAmbireWallet || isArgentWallet || isSmartContract))
+      setIsSmartContractWallet(Boolean(isSafeWallet || isAmbireWallet || isArgentWallet || isSmartContract))
```
Comment (alfetopito) [inline libs/wallet/src/web3-react/hooks/useIsSmartContractWallet.ts:50]: For some reason Safe was not included in this check.
It's covered by the smart contract check, but so are the other wallets 🤷

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts [hunk 4]
```diff
@@ -33,6 +35,11 @@ export function AppDataHooksUpdater(): null {
   const permitData = usePermitDataIfNotAllowance()
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:35]: not related to this PR, so feel free to ignore. 
I was confused at first by this hook when I check the implementation, as this returns the agnostic data only, shouldn't it have the word agnostic on it?

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts [hunk 5]
```diff
@@ -33,6 +35,11 @@ export function AppDataHooksUpdater(): null {
...
   const permitDataPrev = useRef<PermitHookData | undefined>(undefined)
   const hasTradeInfo = !!v2Trade
+  // This is already covered up the dependency chain, but it still slips through some times
+  // Adding this additional check here to try to prevent a race condition to ever allowing this to pass through
+  const isSmartContractWallet = useIsSmartContractWallet()
```
[the remark this answers is not in the document]
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:40]: It's likely not the issue, but technically I guess it could happen?

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts [hunk 6]
```diff
@@ -46,18 +53,18 @@ export function AppDataHooksUpdater(): null {
       preInteractionHooks: permitData ? [permitData] : undefined,
     })
 
-    if (hooks) {
+    if (!isSmartContractWallet && !isNativeSell && hooks) {
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:56]: I guess this is already handled when you get the hook data (u just explained me how the generation return undefined in the case of native sell)

File: apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    return false
+  }
+
+  try {
+    const appData = JSON.parse(fullAppData) as AppDataRootSchema
+
+    return !!appData.metadata?.hooks
+  } catch (e) {
+    // Does not match the schema, use a regex
+    return /"hooks"/.test(fullAppData)
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts:14]: mm, a bit weird check... 
so if the appData has any field with `"hooks"`, then you return true?

what case are you trying to cover? if its us who generated the hook, then it won't be an issue

File: apps/cowswap-frontend/src/modules/swap/services/ethFlow/index.ts [hunk 8]
```diff
@@ -34,6 +37,12 @@ export async function ethFlow(
     return undefined
   }
 
+  if (appDataContainsHooks(orderParamsOriginal.appData.fullAppData)) {
+    reportAppDataWithHooks(orderParamsOriginal)
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/swap/services/ethFlow/index.ts:41]: i guess this should never happen, maybe add a comment so its removed later when we find the real issue. It seems we do the checks multiple times (at least 3 places check for ethFlow and remove the hooks, makes it too hard to reason about).

File: apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleApprovalFlow.ts (hunk 2 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleApprovalFlow.ts:66]: same comment as above

File: apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleEthFlow.ts [hunk 9]
```diff
@@ -75,6 +78,12 @@ export async function safeBundleEthFlow(
       })
     }
 
+    if (appDataContainsHooks(orderParams.appData.fullAppData)) {
+      reportAppDataWithHooks(orderParams)
+      // wipe out the hooks
+      orderParams.appData = await updateHooksOnAppData(orderParams.appData, undefined)
+    }
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/swap/services/safeBundleFlow/safeBundleEthFlow.ts:85]: same :)

File: libs/common-utils/src/sentry.ts [hunk 10]
```diff
@@ -7,3 +7,11 @@ export function reportPermitWithDefaultSigner(params: Record<any, any>): void {
     contexts: { params },
   })
 }
+
+export function reportAppDataWithHooks(params: Record<any, any>): void {
+  // report to sentry if we ever use hooks in the app data
+  Sentry.captureException("Hooks are present in the app data when it shouldn't", {
+    tags: { errorType: 'appDataWithHooks' },
```
Comment (anxolin) [inline libs/common-utils/src/sentry.ts:14]: Do you know if we can get a slack message in this case from sentry, so we see them in real time?

**Review by anxolin — COMMENTED** [review]
Soft approve (to try this out)

Nice work investigating. The flow is a bit hard to follow, but let's see if with these fixes and the sentry we can understand the real cause of the problem

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts (hunk 6 above)
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:56]: Yes it should be.
But it's still getting through.

File: apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts (hunk 7 above)
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts:14]: And the regex is just a precaution in case for some reason we fail to parse the appData.

File: libs/common-utils/src/sentry.ts (hunk 10 above)
Comment (alfetopito) [reply libs/common-utils/src/sentry.ts:14]: Will try to set a sentry alarm for this.

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:40]: I think it's unlikely, but in any case: 01a1e16e6bf292015565bef562ec64f614288cdd

File: apps/cowswap-frontend/src/modules/swap/services/ethFlow/index.ts (hunk 8 above)
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/swap/services/ethFlow/index.ts:41]: 972db4f606ebabe64840e84257e9b8f8ca3e3c3c

File: libs/common-utils/src/sentry.ts (hunk 10 above)
Comment (alfetopito) [reply libs/common-utils/src/sentry.ts:14]: Created alert that'll trigger once a day and notify on slack if it has that error type of the other one https://cowprotocol.sentry.io/alerts/rules/cowswap/14952669/details/

**Comment by elena-zh** [comment]
Hey @alfetopito , I have not been able to reproduce it neither on Prod nor in PR. 
I also tried signing (not executing) with a partial approval, wallets chaging. But no hooks in the appData for me

File: libs/common-utils/src/sentry.ts (hunk 10 above)
Comment (alfetopito) [reply libs/common-utils/src/sentry.ts:14]: @anxolin I triggered it locally to test.
See the event details https://cowprotocol.sentry.io/issues/4918391553/?environment=local&project=5905822&query=permit&statsPeriod=1h&stream_index=0

image

File: apps/cowswap-frontend/src/common/hooks/featureFlags/useIsPermitEnabled.ts [hunk 11]
```diff
@@ -2,5 +2,5 @@ import { useIsSmartContractWallet } from '@cowprotocol/wallet'
 
 export function useIsPermitEnabled(): boolean {
   // Permit is only available for EOAs
-  return !useIsSmartContractWallet()
+  return useIsSmartContractWallet() === false
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/common/hooks/featureFlags/useIsPermitEnabled.ts:5]: Is there any difference?
Just curious

**Review by shoom3301 — APPROVED** [review]

File: apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts (hunk 7 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts:14]: But what i mean, isn't it covered by `const appData = JSON.parse(fullAppData) as AppDataRootSchema`, why would it have some other JSON structure? testing for the presence of "hooks" feels fragile

File: libs/common-utils/src/sentry.ts (hunk 10 above)
Comment (anxolin) [reply libs/common-utils/src/sentry.ts:14]: Awesome, thanks for this

**Review by anxolin — COMMENTED** [review]
Not a fan of this defensive programming, but I don't have a better suggestion as you already made a big effort searching for the root cause. 

Let's see if alerts helps to understand what is happening

File: apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts (hunk 7 above)
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/permit/utils/appDataContainsHooks.ts:14]: Fair, the regex is an overkill.
If the data doesn't match the expected format it'll probably not be executable anyway.
I'll remove this bit.

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:40]: Cool!

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts (hunk 4 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts]: thanks for changing

**Review by anxolin — APPROVED** [review]
